gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,202 @@
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      Predicate<CommitMeta> expiredCommitPredicate) {
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+
+      Set<ContentKey> deletedKeys = new HashSet<>();
+      AtomicBoolean commitHead = new AtomicBoolean(true);
+      commits.forEach(
+          (logEntry) -> {
+            boolean isExpired;
+            if (commitHead.get()) {
+              // If the commit is head commit, then it should be kept live if the table is reachable
+              isExpired = false;
+              commitHead.set(false);
+            } else {
+              // check expiry based on cutoff time
+              isExpired = expiredCommitPredicate.test(logEntry.getCommitMeta());
+            }
+
+            if (logEntry.getOperations() != null) {
+              logEntry
+                  .getOperations()
+                  .forEach(
+                      operation -> {
+                        if (operation instanceof Operation.Delete) {
+                          deletedKeys.add(operation.getKey());
+                        }
+                      });
+            }
+            handleCommit(
+                contentValuesCollector,
+                reference,
+                logEntry.getCommitMeta(),
+                deletedKeys,
+                isExpired);
+          });
+    } catch (RuntimeException e) {
+      if (e.getCause() instanceof NessieNotFoundException) {
+        LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
+      } else {
+        throw e;
+      }
+    } catch (NessieNotFoundException e) {
+      LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
+    }
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void handleCommit(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      CommitMeta commitMeta,
+      Set<ContentKey> deletedKeys,
+      boolean isExpired) {
+    try {
+      Reference commitReference = referenceWithHash(reference, commitMeta);
+      List<ContentKey> allKeys =
+          api.getEntries().reference(commitReference).get().getEntries().stream()
+              .map(EntriesResponse.Entry::getName)
+              .collect(Collectors.toList());
+      api.getContent()
+          .reference(commitReference)
+          .keys(allKeys)
+          .get()
+          .forEach(
+              (contentKey, content) -> {
+                boolean isContentExpired = isExpired;
+                if (deletedKeys.contains(contentKey)) {
+                  // table is dropped. So, expire all the contents for this ContentKey independent
+                  // of cutOffTimeStamp.
+                  isContentExpired = true;
```

Comment:
If a table was dropped in current commit, is it even possible to get the content object for it?

Reply:
yes, when table is dropped it comes as `DELETE` operation commit to database adapter with content key which is deleted.
But the delete reference is missed scenario here as algorithm needs reference to walk. 
So, need to rethink on this. I will keep PR for draft again. we can wait.